File: medium/0658_find_k_closest_elements/solution.cpp
#include <algorithm>
#include <vector>

class Solution {
public:
  std::vector<int> findClosestElements(std::vector<int> &arr, int k, int x) {
    auto left = std::lower_bound(arr.begin(), arr.end(), x);
    auto right = left;

    while (right - left < k) {
      if (left != arr.begin() && right != arr.end()) {
        if (x - *(left - 1) <= *right - x)
          left--;
        else
          right++;
      } else if (left == arr.begin()) {
        right++;
      } else {
        --left;
      }
    }

    std::vector<int> neighbors;

    for (; left != right; ++left)
      neighbors.push_back(*left);

    return neighbors;
  }
};

File: medium/0658_find_k_closest_elements/solution.py
from bisect import bisect_left


class Solution:
    def findClosestElements(self, arr: list[int], k: int, x: int) -> list[int]:
        n = len(arr)

        left = bisect_left(arr, x)
        right = left

        while right - left < k:
            if left != 0 and right != n:
                if x - arr[left - 1] <= arr[right] - x:
                    left -= 1
                else:
                    right += 1
            elif right == n:
                left -= 1
            else:
                right += 1

        neighbors = []

        for i in range(left, right):
            neighbors.append(arr[i])

        return neighbors
